voucher/utils.py

```diff
@@ -39,18 +39,32 @@ def get_eligible_coupon_choices(voucher):
         list of tuple:
             list of ('[<product_id>, <coupon_id>]', <course title>) for an eligible coupon / CourseRun match
     """
-    course_matches = CourseRun.objects.filter(
-        course__readable_id__exact=voucher.course_id_input,
-        course__title__iexact=voucher.course_title_input,
-        start_date__date=voucher.course_start_date_input,
-    )
-    if not course_matches.exists():
+    course_matches = None
+    # Search for an exact match if all inputs exist
+    if voucher.course_id_input and voucher.course_title_input:
+        course_matches = CourseRun.objects.filter(
+            course__readable_id__exact=voucher.course_id_input,
+            course__title__exact=voucher.course_title_input,
+            start_date__date=voucher.course_start_date_input,
+        )
+
+    # Search for partial matches if no exact match was found
+    if course_matches is None or not course_matches.exists():
         # Try partial matching
         course_matches = CourseRun.objects.filter(
-            Q(course__readable_id__exact=voucher.course_id_input)
-            | Q(course__title__iexact=voucher.course_title_input)
+            (
+                Q(course__readable_id__icontains=voucher.course_id_input)
+                if voucher.course_id_input
+                else Q()
+            )
+            | (
+                Q(course__title__icontains=voucher.course_title_input)
```

Comment:
A helper function like this could be used in read_pdf():
```python
def remove_extra_spaces(text):
    """
    Remove extra spaces from text
    
    Args:
        text(str): The text to remove spaces from
        
    Returns:
        str: The text with extra spaces removed
    """
    if text:
        return re.sub(r'\s+', ' ', text.strip())
```